File: sdks/python/apoa/apoa.py
"""Application-facing APOA facade."""

from __future__ import annotations

import re
from datetime import datetime, timedelta, timezone
from typing import Any

from .client import APOAClient, create_client
from .types import (
    APOADefinition,
    APOAToken,
    Agent,
    AgentProvider,
    APIAccessConfig,
    AuthorizationResult,
    BrowserSessionConfig,
    ConstraintMap,
    LegalFramework,
    Principal,
    Rule,
    ServiceAuthorization,
    SigningOptions,
    ValidationOptions,
    ValidationResult,
)


class APOA:
    """Application-facing APOA facade.

    This keeps the protocol-level client intact while giving app developers a
    smaller first path: configure once, then use namespaced resources.
    """

    def __init__(
        self,
        private_key: Any | None = None,
        algorithm: str = "EdDSA",
        revocation_store: Any | None = None,
        audit_store: Any | None = None,
        key_resolver: Any | None = None,
    ) -> None:
        self._client = create_client(
            revocation_store=revocation_store,
            audit_store=audit_store,
            key_resolver=key_resolver,
            default_private_key=private_key,
            default_algorithm=algorithm,
        )
        self.tokens = TokensResource(self._client)
        self.authorizations = AuthorizationsResource(self._client)

    def generate_key_pair(self, algorithm: str | None = None) -> tuple:
        """Generate a key pair using the configured/default algorithm."""
        return self._client.generate_key_pair(algorithm)


class TokensResource:
    """Token operations exposed from the application-facing facade."""

    def __init__(self, client: APOAClient) -> None:
        self._client = client

    def create(
        self,
        definition: APOADefinition,
        options: SigningOptions | None = None,
    ) -> APOAToken:
        return self._client.create_token(definition, options)

    def create_grant(
        self,
        *,
        principal: str | Principal,
        agent: str | Agent,
        service: str | None = None,
        scopes: list[str] | None = None,
        services: list[ServiceAuthorization] | None = None,
        constraints: ConstraintMap | None = None,
        rules: list[Rule] | None = None,
        expires: datetime | str | None = None,
        expires_in: str | None = None,
        revocable: bool = True,
        delegatable: bool = False,
        max_delegation_depth: int | None = None,
        metadata: dict[str, str | int | float | bool | None] | None = None,
        access_mode: str | None = None,
        browser_config: BrowserSessionConfig | None = None,
        api_config: APIAccessConfig | None = None,
        agent_provider: AgentProvider | None = None,
        legal: LegalFramework | None = None,
        options: SigningOptions | None = None,
    ) -> APOAToken:
        definition = _normalize_grant_input(
            principal=principal,
            agent=agent,
            service=service,
            scopes=scopes,
            services=services,
            constraints=constraints,
            rules=rules,
            expires=expires,
            expires_in=expires_in,
            revocable=revocable,
            delegatable=delegatable,
            max_delegation_depth=max_delegation_depth,
            metadata=metadata,
            access_mode=access_mode,
            browser_config=browser_config,
            api_config=api_config,
            agent_provider=agent_provider,
            legal=legal,
        )
        return self._client.create_token(definition, options)

    def validate(
        self,
        token: str | APOAToken,
        options: ValidationOptions | None = None,
        **kwargs: Any,
    ) -> ValidationResult:
        if options is not None and kwargs:
            raise ValueError("Pass either options or validation keyword arguments, not both")
        return self._client.validate_token(token, options or ValidationOptions(**kwargs))

    def parse(self, input_str: str, format: str = "json") -> APOADefinition:
        return self._client.parse_definition(input_str, format)


class AuthorizationsResource:
    """Authorization operations exposed from the application-facing facade."""

    def __init__(self, client: APOAClient) -> None:
        self._client = client

    def check(self, token: APOAToken, service: str, action: str) -> AuthorizationResult:
        return self._client.authorize(token, service, action)


def _normalize_grant_input(
    *,
    principal: str | Principal,
    agent: str | Agent,
    service: str | None,
    scopes: list[str] | None,
    services: list[ServiceAuthorization] | None,
    constraints: ConstraintMap | None,
    rules: list[Rule] | None,
    expires: datetime | str | None,
    expires_in: str | None,
    revocable: bool,
    delegatable: bool,
    max_delegation_depth: int | None,
    metadata: dict[str, str | int | float | bool | None] | None,
    access_mode: str | None,
    browser_config: BrowserSessionConfig | None,
    api_config: APIAccessConfig | None,
    agent_provider: AgentProvider | None,
    legal: LegalFramework | None,
) -> APOADefinition:
    errors: list[str] = []
    normalized_principal = _normalize_principal(principal, errors)
    normalized_agent = _normalize_agent(agent, errors)
    normalized_services = _normalize_services(
        service=service,
        scopes=scopes,
        services=services,
        constraints=constraints,
        access_mode=access_mode,
        browser_config=browser_config,
        api_config=api_config,
        errors=errors,
    )
    normalized_expires = _normalize_expires(expires, expires_in, errors)

    if errors or normalized_principal is None or normalized_agent is None or normalized_expires is None:
        raise _invalid_grant_input(errors)

    return APOADefinition(
        principal=normalized_principal,
        agent=normalized_agent,
        services=normalized_services,
        expires=normalized_expires,
        agent_provider=agent_provider,
        rules=rules,
        revocable=revocable,
        delegatable=delegatable,
        max_delegation_depth=max_delegation_depth,
        metadata=metadata,
        legal=legal,
    )


def _normalize_principal(principal: str | Principal, errors: list[str]) -> Principal | None:
    if isinstance(principal, str) and principal.strip():
        return Principal(id=principal.strip())
    if isinstance(principal, Principal) and principal.id:
        return principal
    errors.append("principal is required; pass a DID string or Principal(id=...)")
    return None


def _normalize_agent(agent: str | Agent, errors: list[str]) -> Agent | None:
    if isinstance(agent, str) and agent.strip():
        return Agent(id=agent.strip())
    if isinstance(agent, Agent) and agent.id:
        return agent
    errors.append("agent is required; pass a DID string or Agent(id=...)")
    return None


def _normalize_services(
    *,
    service: str | None,
    scopes: list[str] | None,
    services: list[ServiceAuthorization] | None,
    constraints: ConstraintMap | None,
    access_mode: str | None,
    browser_config: BrowserSessionConfig | None,
    api_config: APIAccessConfig | None,
    errors: list[str],
) -> list[ServiceAuthorization]:
    if services is not None:
        if len(services) == 0:
            errors.append("services must be a non-empty list when provided")
            return []
        return services

    if not service:
        errors.append("service is required unless services is provided")
        return []
    if not scopes:
        errors.append("scopes must be a non-empty list unless services is provided")
        return []

    return [
        ServiceAuthorization(
            service=service,
            scopes=scopes,
            constraints=constraints,
            access_mode=access_mode,
            browser_config=browser_config,
            api_config=api_config,
        )
    ]


def _normalize_expires(
    expires: datetime | str | None,
    expires_in: str | None,
    errors: list[str],
) -> datetime | str | None:
    if expires is not None and expires_in is not None:
        errors.append("pass either expires or expires_in, not both")
        return None
    if expires is not None:
        return expires
    if expires_in is not None:
        return _parse_duration_from_now(expires_in)
    errors.append("expires or expires_in is required")
    return None


def _parse_duration_from_now(duration: str) -> datetime:
    match = re.fullmatch(r"(\d+)([smhd])", duration)
    if not match:
        raise _invalid_grant_input([
            "expires_in must use a clear duration like '15m', '2h', or '30d'"
        ])

    amount = int(match.group(1))
    if amount <= 0:
        raise _invalid_grant_input(["expires_in duration must be a positive integer"])

    units = {
        "s": timedelta(seconds=amount),
        "m": timedelta(minutes=amount),
        "h": timedelta(hours=amount),
        "d": timedelta(days=amount),
    }
    return datetime.now(timezone.utc) + units[match.group(2)]


def _invalid_grant_input(errors: list[str]) -> ValueError:
    return ValueError(
        "\n".join(
            [
                "Invalid APOA grant input.",
                *(f"- {error}" for error in errors),
                "Minimal shape: principal, agent, service, scopes, expires_in",
            ]
        )
    )

File: sdks/typescript/src/apoa.ts
import { createClient } from './client.js';
import type {
  APOAClient,
  APOADefinition,
  APOAOptions,
  APOAToken,
  AuthorizationResult,
  AuthorizeOptions,
  Principal,
  Agent,
  ServiceAuthorization,
  SimpleGrantInput,
  SigningOptions,
  ValidationOptions,
  ValidationResult,
} from './types.js';

/**
 * Application-facing APOA facade.
 *
 * This keeps the protocol-level APIs intact while giving app developers a
 * smaller first path: configure once, then use namespaced resources.
 */
export class APOA {
  private readonly client: APOAClient;

  readonly tokens: {
    create: (definition: APOADefinition, options?: SigningOptions) => Promise<APOAToken>;
    createGrant: (input: SimpleGrantInput, options?: SigningOptions) => Promise<APOAToken>;
    validate: (
      token: string | APOAToken,
      options?: Omit<ValidationOptions, 'revocationStore'>
    ) => Promise<ValidationResult>;
    parse: (input: string, format?: 'yaml' | 'json') => APOADefinition;
  };

  readonly authorizations: {
    check: (
      token: APOAToken,
      service: string,
      action: string,
      options?: Omit<AuthorizeOptions, 'revocationStore' | 'auditStore'>
    ) => Promise<AuthorizationResult>;
  };

  constructor(options: APOAOptions = {}) {
    const { privateKey, algorithm, kid, ...clientOptions } = options;

    this.client = createClient({
      ...clientOptions,
      defaultSigningOptions: privateKey
        ? { privateKey, algorithm, kid }
        : undefined,
    });

    this.tokens = {
      create: (definition, signingOptions) =>
        this.client.createToken(definition, signingOptions),
      createGrant: async (input, signingOptions) =>
        this.client.createToken(normalizeGrantInput(input), signingOptions),
      validate: (token, validationOptions) =>
        this.client.validateToken(token, validationOptions),
      parse: (input, format) =>
        this.client.parseDefinition(input, format),
    };

    this.authorizations = {
      check: (token, service, action, authorizeOptions) =>
        this.client.authorize(token, service, action, authorizeOptions),
    };
  }

  async generateKeyPair(algorithm?: 'EdDSA' | 'ES256'): Promise<CryptoKeyPair> {
    return this.client.generateKeyPair(algorithm);
  }
}

function normalizeGrantInput(input: SimpleGrantInput): APOADefinition {
  const errors: string[] = [];

  if (!input || typeof input !== 'object') {
    throw invalidGrantInput(['input must be an object']);
  }

  const principal = normalizePrincipal(input.principal, errors);
  const agent = normalizeAgent(input.agent, errors);
  const services = normalizeServices(input, errors);
  const expires = normalizeExpires(input, errors);

  if (errors.length > 0 || !principal || !agent || services.length === 0 || !expires) {
    throw invalidGrantInput(errors);
  }

  return {
    principal,
    agent,
    services,
    expires,
    ...(input.rules ? { rules: input.rules } : {}),
    ...(input.revocable !== undefined ? { revocable: input.revocable } : {}),
    ...(input.delegatable !== undefined ? { delegatable: input.delegatable } : {}),
    ...(input.maxDelegationDepth !== undefined
      ? { maxDelegationDepth: input.maxDelegationDepth }
      : {}),
    ...(input.metadata ? { metadata: input.metadata } : {}),
    ...(input.agentProvider ? { agentProvider: input.agentProvider } : {}),
    ...(input.legal ? { legal: input.legal } : {}),
  };
}

function normalizePrincipal(
  principal: SimpleGrantInput['principal'],
  errors: string[]
): Principal | undefined {
  if (typeof principal === 'string' && principal.trim()) {
    return { id: principal.trim() };
  }
  if (principal && typeof principal === 'object' && principal.id) {
    return principal;
  }
  errors.push('principal is required; pass a DID string or { id }');
  return undefined;
}

function normalizeAgent(
  agent: SimpleGrantInput['agent'],
  errors: string[]
): Agent | undefined {
  if (typeof agent === 'string' && agent.trim()) {
    return { id: agent.trim() };
  }
  if (agent && typeof agent === 'object' && agent.id) {
    return agent;
  }
  errors.push('agent is required; pass a DID string or { id }');
  return undefined;
}

function normalizeServices(
  input: SimpleGrantInput,
  errors: string[]
): ServiceAuthorization[] {
  if (input.services) {
    if (!Array.isArray(input.services) || input.services.length === 0) {
      errors.push('services must be a non-empty array when provided');
      return [];
    }
    return input.services;
  }

  if (!input.service) {
    errors.push('service is required unless services is provided');
    return [];
  }
  if (!input.scopes || !Array.isArray(input.scopes) || input.scopes.length === 0) {
    errors.push('scopes must be a non-empty array unless services is provided');
    return [];
  }

  return [{
    service: input.service,
    scopes: input.scopes,
    ...(input.constraints ? { constraints: input.constraints } : {}),
    ...(input.accessMode ? { accessMode: input.accessMode } : {}),
    ...(input.browserConfig ? { browserConfig: input.browserConfig } : {}),
    ...(input.apiConfig ? { apiConfig: input.apiConfig } : {}),
  }];
}

function normalizeExpires(
  input: SimpleGrantInput,
  errors: string[]
): Date | string | undefined {
  if (input.expires && input.expiresIn) {
    errors.push('pass either expires or expiresIn, not both');
    return undefined;
  }
  if (input.expires) {
    return input.expires;
  }
  if (input.expiresIn) {
    return parseDurationFromNow(input.expiresIn);
  }
  errors.push('expires or expiresIn is required');
  return undefined;
}

function parseDurationFromNow(duration: string): Date {
  const match = /^(\d+)([smhd])$/.exec(duration);
  if (!match) {
    throw invalidGrantInput([
      `expiresIn must use a clear duration like '15m', '2h', or '30d'`,
    ]);
  }

  const amount = Number(match[1]);
  if (!Number.isSafeInteger(amount) || amount <= 0) {
    throw invalidGrantInput(['expiresIn duration must be a positive integer']);
  }

  const unitMs: Record<string, number> = {
    s: 1000,
    m: 60 * 1000,
    h: 60 * 60 * 1000,
    d: 24 * 60 * 60 * 1000,
  };

  return new Date(Date.now() + amount * unitMs[match[2]]);
}

function invalidGrantInput(errors: string[]): Error {
  return new Error(
    [
      'Invalid APOA grant input.',
      ...errors.map((error) => `- ${error}`),
      'Minimal shape: { principal, agent, service, scopes, expiresIn }',
    ].join('\n')
  );
}
